src/diagnostics.rs

```diff
@@ -0,0 +1,204 @@
+        };
+        let file_map = Rc::new(FileMap::new(source, markdown));
+
+        Diagnostics { file_map, out: Diagnostics::create_out_stream() }
+    }
+
+    pub fn first_line(&self, range: &Range<usize>) -> Range<usize> {
+        let start =
+            Span::from_offset(self.file_map.span().start(), ByteOffset(range.start as i64)).end();
+        let line = self.file_map.location(start).unwrap().0;
+        let line_span = self.file_map.line_span(line).unwrap();
+        // get rid of newline
+        let len = self.file_map.src_slice(line_span).unwrap().trim_end().len();
+        Range { start: range.start, end: range.start + len }
+    }
+
+    fn diagnostic(&mut self, severity: Severity, message: String) -> DiagnosticBuilder<'a, '_> {
+        DiagnosticBuilder {
+            file_map: &self.file_map,
+            out: &mut self.out,
+            diagnostics: Vec::new(),
+            severity,
+            message,
+            code: None,
+            labels: Vec::new(),
+        }
+    }
+
+    pub fn bug<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        let mut diag =
+            Some(self.diagnostic(Severity::Bug, message.into()).note("please report this"));
+        backtrace::trace(|frame| {
+            let ip = frame.ip();
+            backtrace::resolve(ip, |symbol| {
+                diag = Some(diag.take().unwrap().note(format!(
+                    "in heradoc file {:?} name {:?} line {:?} address {:?}",
+                    symbol.filename(),
+                    symbol.name(),
+                    symbol.lineno(),
+                    symbol.addr()
+                )));
+            });
+            true
+        });
+        diag.unwrap()
+    }
+
+    pub fn error<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Error, message.into())
+    }
+
+    pub fn warning<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Warning, message.into())
+    }
+
+    pub fn note<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Note, message.into())
+    }
+
+    pub fn help<S: Into<String>>(&mut self, message: S) -> DiagnosticBuilder<'a, '_> {
+        self.diagnostic(Severity::Help, message.into())
+    }
+}
+
+#[must_use = "call `emit` to emit the diagnostic"]
+pub struct DiagnosticBuilder<'a: 'b, 'b> {
+    file_map: &'b FileMap<&'a str>,
+    out: &'b mut StandardStream,
+    diagnostics: Vec<Diagnostic>,
+
+    severity: Severity,
+    message: String,
+    code: Option<String>,
+    labels: Vec<Label>,
+}
+
+impl<'a: 'b, 'b> DiagnosticBuilder<'a, 'b> {
+    pub fn emit(self) {
+        let Self { file_map, out, mut diagnostics, severity, message, code, labels } = self;
+        diagnostics.push(Diagnostic { severity, message, code, labels });
+
+        // ignore output errors, because where would we log them anyway?!
```

Comment:
If we panic, that'll still try to write a message to the output, which is already gone. So the only sane way to handle this would be an `abort()`, but then the user might wonder why the program is just gone without any error message.

So there isn't really any sane way to handle this. This is something that just shouldn't happen :) I guess I should just panic here and let the `panic_handler` take care of the rest. Anyway, even this might be bad, because the panic handler writes stuff again.